### PR TITLE
Introduce foundations of APIs

### PR DESCRIPTION
**Pull request by enescakir**
This PR introduces foundations of APIs in clover.

We use Roda as web framework. It supports running nested roda apps from
a parent roda app. I divided roda app to two sub roda apps as CloverWeb
and CloverApi to isolate web console and API endpoints. It helps to
enable only necessary plugins and helper function for sub apps. For
example CloverApi doesn't need HTML rendering plugin.

- Move old Clover < Roda class to `clover_web.rb`
- Create new sub roda app for API `clover_api.rb`
- Put common roda app statements to `routes\base.rb`

Currently, API uses Rodauth's JWT plugin, I need to work more on
authentication part.

## API Paths
We focused on 3 different path formats.
- vm/{VM_ID}
- project/{PROJECT_ID}/vm/{VM_NAME}
- project/{PROJECT_ID}/location/{LOCATION_NAME}/vm/{VM_NAME}

Project prefix helps to isolate different projects' resources and
prevents to do mistake on wrong project. Also Vm names aren't globally
unique, so with project prefix we can use vm names in paths.

Location prefix is an investment for future. It will help us to
determine maiden location of a resource.

## API Reference
Documentation: https://documenter.getpostman.com/view/2986669/2s93zFWJnS

- `GET    /api/project`
- `POST   /api/project`
```json
{
    "name": "PROJECT_NAME"
}
```
- `GET    /api/project/{PROJECT_ID}`
- `GET    /api/project/{PROJECT_ID}/location/{LOCATION_NAME}/vm`
- `POST   /api/project/{PROJECT_ID}/location/{LOCATION_NAME}/vm`
```json
{
    "name": "",
    "unix_user": "",
    "public_key": "",
    "size": "",
    "boot_image": ""
}
```
- `GET    /api/project/{PROJECT_ID}/location/{LOCATION_NAME}/vm/{VM_NAME}`
- `DELETE /api/project/{PROJECT_ID}/location/{LOCATION_NAME}/vm/{VM_NAME}`

## curl Examples
```
BASE_URL="localhost:9292/api"
USER_NAME=""
USER_PASSWORD=""
PROJECT_ID=""
LOCATION="hetzner-hel1"
VM_NAME="test-vm"

# Login
curl  "$BASE_URL/login?login=$USER_NAME&password=$USER_PASSWORD" \
    -s -I -X POST \
    -H "Content-Type: application/json"

JWT_TOKEN="<FROM_LOGIN_REQUEST>"

# List project
curl "$BASE_URL/project" \
    -H "Authorization: Bearer $JWT_TOKEN"

# Create a VM
curl "$BASE_URL/project/$PROJECT_ID/location/$LOCATION/vm" \
    -X POST \
    -H "Authorization: Bearer $JWT_TOKEN" \
    -H "Content-Type: application/json" \
    --data '{
        "name": "test-vm",
        "unix_user": "ubi",
        "public_key": "ssh-ed25519 ABCDEF",
        "size": "c5a.2x",
        "boot_image": "ubuntu-jammy"
    }'

# Get VM
curl "$BASE_URL/project/$PROJECT_ID/location/$LOCATION/vm/$VM_NAME" \
    -H "Authorization: Bearer $JWT_TOKEN"
```

### CODE REVIEW

File: clover_api.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  plugin :error_handler do |e|
+    error = parse_error(e)
+
+    {error: error}.to_json
+  end
+
+  plugin :rodauth do
+    # Use JWT token for now, think more about authentication.
```
Comment (fdr) [inline clover_api.rb:36]:
```suggestion
```

I think the question of fancier authentication will be on our minds enough that we don't need a tiny comment to remind us. Mostly the upshot of such a design would be signing the content of the request and checking signatures of responses, which does oblige clients to be somewhat more complex. We may have to scope that and other more-stringent authentication stuff to particular projects, in practice I've found the ratio of development/staging projects to production projects is high...and only the production projects can justify significantly more annoyance and paranoia. It'd be nice if we were cognizant of this and avoided the need for a second identity (e.g. myname+prod@example.com), something I've done quite often.

File: lib/authorization.rb [hunk 2]
```diff
@@ -51,7 +51,10 @@ def self.generate_default_acls(subject, object)
 
   module Dataset
     def authorized(subject_id, actions)
-      where(id: Authorization.authorized_resources(subject_id, actions))
+      # We can't use "id" column directly, because it's ambiguous in big joined queries.
+      # We need to determine table of id explicitly.
+      from = @opts[:from].first
```
Comment (fdr) [inline lib/authorization.rb:56]: sneaky, where does `@opts` come from? I imagine it's make-sense thing in sequel, but help out the reader (me) a bit if it's short enough (if it's not, a link is better)

File: routes/api/project/location/vm.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+require "ulid"
+
+class CloverApi
+  hash_branch(:project_location_prefix, "vm") do |r|
+    @serializer = Serializers::Api::Vm
+
+    r.get true do
+      vms = @project.vms_dataset.where(location: @location).authorized(@current_user.id, "Vm:view").all
```
Comment (fdr) [inline routes/api/project/location/vm.rb:10]: `#authorized` is so cool. Perhaps it can be made even less error prone, but it's so powerful.

File: routes/base.rb [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      end
+    when :test
+      ::Mail.defaults do
+        delivery_method :test
+      end
+    end
+    # :nocov:
+  end
+
+  # TODO(enes): Handle exception better way, maybe bind http codes with exceptions
```
Comment (fdr) [inline routes/base.rb:46]:
```suggestion
  # Assign some HTTP response codes to common exceptions.
```

I think this TODO so terse we'll (or at least many of us) wonder what the heck we were talking about in a few months. I think a reasonable software designer can come up with the idea of in-situ exceptions-with-http-codes on their own (although they may wonder if it's a good idea, to which I think, with some experience, I would say: yeah probably).

If you really want to emphasize the point, in a follow-up, find some spots where you want a nice error message and then implement (and feel out) the pattern at your earliest convenience. Sometimes to show is really better to describe.

**Review by fdr — APPROVED** [review]
I'm a big fan of how the authorization system is composing with model queries.

**Comment by fdr** [comment]
Also thanks for the detailed commit message, though I think it might be useful how you interact with the API to iterate with it, e.g. if that's "curl" or some ruby expressions or what.

It made me think about the "project/location/$location" thing being a bit wordy, but after thinking things over I felt you might have made the best choice.

File: routes/api/project/location/vm.rb (hunk 3 above)
Comment (enescakir) [reply routes/api/project/location/vm.rb:10]: All the time, we give `@current_user.id` as first parameter in `routes`. I couldn't find a way to pass this value in a hidden way.

Hardcoded powers are at everywhere, they feels like open to typos and errors.

File: clover_api.rb (hunk 1 above)
Comment (enescakir) [reply clover_api.rb]: I think we can develop project based some features

File: lib/authorization.rb (hunk 2 above)
Comment (enescakir) [reply lib/authorization.rb:57]: I added comment

File: routes/base.rb (hunk 4 above)
Comment (enescakir) [reply routes/base.rb]: I updated comment. I will show some example in future probably

**Comment by enescakir** [comment]
I added a Postman documentation link to commit message: https://documenter.getpostman.com/view/2986669/2s93zFWJnS
Screenshot 2023-07-07 at 14 18 22
It has examples for different ways.
Currently our UI doesn't have a way to generate JWT tokens. I will add it. 
You can use built-in `/api/login` to get JWT token. Documentation has it.

**Comment by enescakir** [comment]
I created a draft PR for tag changes. We can discuss about tags more
https://github.com/ubicloud/ubicloud/pull/265

**Comment by enescakir** [comment]
I added a few curl examples to commit message

------
## curl Examples
```
BASE_URL="localhost:9292/api"
USER_NAME=""
USER_PASSWORD=""
PROJECT_ID=""
LOCATION="hetzner-hel1"
VM_NAME="test-vm"

# Login
curl  "$BASE_URL/login?login=$USER_NAME&password=$USER_PASSWORD" \
    -s -I -X POST \
    -H "Content-Type: application/json"

JWT_TOKEN="<FROM_LOGIN_REQUEST>"

# List project
curl "$BASE_URL/project" \
    -H "Authorization: Bearer $JWT_TOKEN"

# Create a VM
curl "$BASE_URL/project/$PROJECT_ID/location/$LOCATION/vm" \
    -X POST \
    -H "Authorization: Bearer $JWT_TOKEN" \
    -H "Content-Type: application/json" \
    --data '{
        "name": "test-vm",
        "unix_user": "ubi",
        "public_key": "ssh-ed25519 ABCDEF",
        "size": "c5a.2x",
        "boot_image": "ubuntu-jammy"
    }'

# Get VM
curl "$BASE_URL/project/$PROJECT_ID/location/$LOCATION/vm/$VM_NAME" \
    -H "Authorization: Bearer $JWT_TOKEN"
```
